compiler/src/main/kotlin/dev/zacsweers/metro/compiler/ir/graph/sharding/ShardingOrchestrator.kt

```diff
@@ -0,0 +1,114 @@
+// Copyright (C) 2025 Zac Sweers
+// SPDX-License-Identifier: Apache-2.0
+package dev.zacsweers.metro.compiler.ir.graph.sharding
+
+import dev.zacsweers.metro.compiler.MetroOptions
+import dev.zacsweers.metro.compiler.graph.Component
+import dev.zacsweers.metro.compiler.graph.TopoSortResult
+import dev.zacsweers.metro.compiler.ir.IrTypeKey
+import dev.zacsweers.metro.compiler.ir.graph.DependencyGraphNode
+import dev.zacsweers.metro.compiler.isInvisibleGeneratedGraph
+
+/**
+ * Partitions bindings into shards when a graph exceeds [MetroOptions.keysPerGraphShard].
+ *
+ * Keeps SCCs together to preserve circular dependencies (including Provider<T> cycles), respects
+ * topological ordering, and merges small tail shards.
+ *
+ * Returns null if sharding isn't needed:
+ * - Graph sharding is disabled
+ * - There are too few bindings
+ * - It's either a dynamic graph or a graph extension
+ */
+internal class ShardingOrchestrator(
+  private val node: DependencyGraphNode,
+  private val options: MetroOptions,
+) {
+
+  /**
+   * Returns shard groups in topologically sorted order, or null if sharding isn't needed.
+   * Each list contains type keys for the bindings in that shard.
+   */
+  fun computeShardGroups(topologyData: TopoSortResult<IrTypeKey>?): List<List<IrTypeKey>>? {
```

Comment:
Seems we should just never call this if we have a null result? This is something I see AI do often and I'd like to avoid it since it creates ambiguous signatures (does it return null only if the input is null? etc)